ostd/src/arch/x86/iommu/fault.rs

```diff
@@ -201,20 +210,61 @@ bitflags! {
     }
 }
 
-pub(super) static FAULT_EVENT_REGS: Once<FaultEventRegisters> = Once::new();
+pub(super) static FAULT_EVENT_REGS: Once<SpinLock<FaultEventRegisters, LocalIrqDisabled>> =
+    Once::new();
 
 /// Initializes the fault reporting function.
 ///
 /// # Safety
 ///
 /// User must ensure the base_register_vaddr is read from DRHD
 pub(super) unsafe fn init(base_register_vaddr: Vaddr) {
-    FAULT_EVENT_REGS.call_once(|| FaultEventRegisters::new(base_register_vaddr));
+    FAULT_EVENT_REGS.call_once(|| SpinLock::new(FaultEventRegisters::new(base_register_vaddr)));
+}
+
+fn iommu_fault_handler(frame: &TrapFrame) {
+    let mut fault_event_regs = FAULT_EVENT_REGS.get().unwrap().lock();
+
+    primary_fault_handler(&mut fault_event_regs);
+
+    let fault_status = fault_event_regs.status();
+    if fault_status.contains(FaultStatus::IQE | FaultStatus::ICE | FaultStatus::ITE) {
+        panic!(
+            "Catch IOMMU invalidation error. Fault status: {:x?}",
+            fault_status
+        );
+    }
 }
 
-fn iommu_page_fault_handler(frame: &TrapFrame) {
-    let fault_event = FAULT_EVENT_REGS.get().unwrap();
-    let index = (fault_event.status().bits & FaultStatus::FRI.bits) >> 8;
-    let recording = FaultRecording(fault_event.recordings[index as usize].read());
-    info!("Catch iommu page fault, recording:{:x?}", recording)
+fn primary_fault_handler(
+    fault_event_regs: &mut SpinLockGuard<FaultEventRegisters, LocalIrqDisabled>,
+) {
+    let mut fault_status = fault_event_regs.status();
+    if !fault_status.contains(FaultStatus::PPF) {
+        return;
+    }
+
+    let start_index = ((fault_event_regs.status().bits & FaultStatus::FRI.bits) >> 8) as usize;
+    for index in start_index..fault_event_regs.recordings.len() {
```

Comment:
This loop can be simplified to:
```rust
for recording in fault_event_regs.recordings.iter_mut().advance_by(start_index) {
  //....
}
```
By doing this, you eliminate the need to use `fault_event_regs.recordings[index]`, which also performs a bounds check.